### ISSUE TITLE
Courier: Information regarding media item is missing in database after transfer

### ISSUE THREAD

**Issue by mupace** [umbraco/Umbraco.Courier.Issues#14]
I am using umbraco v7.15.3 and courier v3.1.8. What I see is courier does not insert information for file path in cmsMedia table (which came with umbraco v7.8). This happens with any type of media upload dialog. (Upload, Image Processor Upload)

Media Node and Resource (image or file) gets transferred. However having information in cmsMedia table is missing makes media pickers treats media items as folders. You cannot pick any of them at target anymore and media picker browser tries to show children items of media item.

This might be related to previous issue, which I am also having regarding cache for media files not being updated at the end of this delivery. #13 

Reproduction
------------

If you're filing a bug, please describe how to reproduce it. Include as much
relevant information as possible, such as:

### Bug summary

This happens even you try to send a single media item with any type of file (PDF, jpg you name it). 

### Specifics
Umbraco Version: 7.15.3
Courier Version: 3.1.8

### Steps to reproduce

1. Head to media tree in Umbraco backoffice 
2. Create a new media item and upload a file
3. Send media item with Courier to destination
4. Check destination backoffice to see if media item delivered 
5. Head to destination database, query cmsMedia table if any information insertted
6. Head to Content tree on destination and try to use a media picker to pick new media item.

### Expected result

1. We should see media item in the destination backoffice
2. We should have information in cmsMedia table and also it should have been able to be used by media pickers.

### Actual result

1. Only the media item is there
2. No information in cmsMedia table
3. Cannot pick item with media pickers

---
_This item has been added to our backlog [AB#7633](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/7633)_

**Comment by mjbarlow** [comment]
This is an issue that we are also facing. Is there an update on this?

**Comment by TimGeyssens** [comment]
Doesn't seem to get a lot of love... (if any) https://github.com/umbraco/Umbraco.Courier.Issues/issues/1

Might be best to shoot an email or log into zendesk to support...

**Comment by TimGeyssens** [comment]
Latest update is from 16/10/2019 https://our.umbraco.com/packages/umbraco-pro/umbraco-courier/, been in touch a bit to find out about Forms support, basicly they say they don't provide a warranty for non severity 1 bugs... https://umbraco.com/umbraco-cms-pricing/severity-levels/ for Courier this means 'Courier defect preventing publishing of time-critical content without a workaround." but from what I read this issue is kinda preventing publishing...

**Comment by TimGeyssens** [comment]
So the "Umbraco agrees to provide a fix that corrects or mitigates the behavior introduced by the bug within 7 business days from the date the bug is confirmed by mutual agreement." 
https://umbraco.com/umbraco-cms-pricing/bug-fix-warranty/
Doesn't seem to be respected here... unless HQ sees this bug not as severity 1

**Comment by TimGeyssens** [comment]
But of course I think you should upgrade to latest v of both umbraco 7 and courier... To make sure it isn't a bug that is already fixed

**Comment by mupace** [comment]
Hey @TimGeyssens ,

Actually v3.1.8 is the latest and 7.15.3 is just 1 security patch behind from latest, when I created this ticket 7.15.4 wasn't out. So I don't think it will make any difference at this point. 

On the other hand we won't be moving to 8.x anytime soon due to business reasons. So I guess somewhat I will need to fix this myself by overriding a data resolver or sth. Looks like it is time to get our hands dirty...

**Comment by TimGeyssens** [comment]
True, then I would suggest getting in touch with support... since personally I think this is a severity 1 bug... (are you just using the build in media picker?)

**Comment by mupace** [comment]
Yep. Built-in media picker, built-in image processor.

**Comment by mjbarlow** [comment]
Agree, this basically makes it pointless of couriering media items, as you can't pick them afterwards.

**Comment by mjbarlow** [comment]
This is how couriered images now look, this should show the image thumbnail.
<img width="278" alt="courier-image-issue" src="https://user-images.githubusercontent.com/1984515/77860053-466e5a80-720d-11ea-8104-6d6b22ace4fb.png">

**Comment by TimGeyssens** [comment]
Shot the q on twitter: https://twitter.com/timgeyssens/status/1244344665719046144 but send it to support and say that you consider this a pretty severe bug that renders courier useless...

**Comment by TimGeyssens** [comment]
So anyone got in touch with support? Since here there doesn't seem to be any official response...

**Comment by mjbarlow** [comment]
I've asked my client to get in touch with them referencing this issue via the paid support channel (Courier licence) to ask for resolution. Thanks for pushing this, yes would be great to have some kind of official response here...

**Comment by TimGeyssens** [comment]
Sweet, yeah hit the same with a Contour bug, so having the conv in the open helps getting some attention to it :)

**Comment by TimGeyssens** [comment]
@mjbarlow any update?

**Comment by mjbarlow** [comment]
My client responded with this, this was the reply from Umbraco Courier support:

"I have now talked with my team, and we are still looking into the case and for now, there are no further updates right now.
The best you can do to get the latest update is to keep an eye of the issue on the issue tracker https://github.com/umbraco/Umbraco.Courier.Issues/issues/14

The issue is not classified under severity 1 when it's a severity 1, then it means the bug is a security issue or loss of data.
Currently, we don´t have an ETA on when this will be fixed and ready in a new release of Courier. "

**Comment by mjbarlow** [comment]
I agree with @TimGeyssens here, this should be categorized as severity 1, as it effects editing and publishing on the destination server,

**Comment by TimGeyssens** [comment]
If this was a physical product you would pretty much ask for a refund...not sure what the laws are around selling software that doesn't deliver what it promises... I would say take your business to uSync https://jumoo.co.uk/products.php ...

**Comment by mjbarlow** [comment]
Good idea, will suggest the uSync route, as this doesn't appear to be a priority, and it's effecting my client commercially. Thanks @TimGeyssens and thanks for pushing this!

**Comment by gordonsaxby** [comment]
This is not what I had hoped to find ... I have this exact same problem with Umbraco 7.15.5 and Courier 3.1.8

There seems to be little point in having / using Courier if it can't do the simple task of copying images!?

**Comment by TimGeyssens** [comment]
Refund....

**Comment by mupace** [comment]
I am sorry that I haven't included this here. 

I've sent a message to Umbraco support and the answer was a little bit puzzling. It says that they "might be replacing" courier with umbraco deploy (as in umbraco cloud) in future and they are too busy with current dev at hand - which is why they might not fix this. 

We have planned to create a custom resolver to address this issue but it has never happened.

**Comment by gordonsaxby** [comment]
> We have planned to create a custom resolver to address this issue but it has never happened.

Did you have a plan in mind of how a custom resolver would / could have fixed the issue?

**Comment by mupace** [comment]
Courier has specific resolver for each property/data type. Normally you need to create one when you create a custom type with a custom property editor. This is how courier understands what is in the data. For example, if property has ability to store a content id, courier checks and adds targeted content as dependency to the node. 

So I believe a resolver for Image Cropper Upload can address this issue. However, creating one (since this is a specific case) requires back and forth to set up test case - in another words it can take quite some time.

**Comment by gordonsaxby** [comment]
I appreciate that you can create resolvers for Courier (currently working on some), but the question is how to fix the actual problem? Would it be a case of resorting to raw SQL commands to update the cmsMedia table?

**Comment by mupace** [comment]
I believe yes; sadly.

**Comment by gordonsaxby** [comment]
I'm not sure it's possible - or at least, I have found it yet. During "ExtractingProperty" for a resolver on "Umbraco.ImageCropper", I can get the node Id and the mediaPath for adding a cmsMedia record. Unfortunately, I cannot find out how to get versionId.

The first time the image is created, the cmsPropertyData record, which contains the versionId does not yet exist. I assume the entire Courier process is within a transaction?

The code I have would work the second time an image was Courier'd as the cmsPropertyData record for it ("umbracoFile") would then exist.

Assuming there isn't a way of getting the versionId ... I am scuppered :-(

**Comment by gordonsaxby** [comment]
Well, a miracle happened! I came across this post which gave me the basics for a solution

https://our.umbraco.com/forum/umbraco-courier/85114-deploying-media-with-courier-and-umbracofilesystemprovidersazure

In the "Execute" of "MediaEventQueueListener" (although I renamed mine), I checked the database for the existance of a cmsMedia record, for the Media item. If it doesn't exist, I get the csmPropertyData record, extract the Version Id and media path and create a cmsMedia record. Seems to work!?
